design/kanister-kopia-integration.md

```diff
@@ -0,0 +1,471 @@
+- Accessing the server requires the service name, a server username, and a password without any knowledge of the backend storage location.
+- To authorize access, a list of server usernames and passwords must be added prior to starting the server.
+- The server also uses TLS certificates to secure incoming connections to it.
+- Kanister users can configure a repository via a newly added Custom Resource called `RepositoryServer` as described in the following section.
+
+#### Custom Resource Definition
+
+This design proposes a new Custom Resource Definition (CRD) named `RepositoryServer`, to represent a Kopia repository server.
+It is a namespace-scoped resource, owned and managed by a new `RepositoryServerController` controller in the `kanister` namespace.
+As mentioned above, the CRD offers a set of parameters to configure a Kopia repository server.
+
+To limit the new controller's RBAC scope, all the `RepositoryServer` resources are created in the `kanister` namespace.
+
+A sample `RepositoryServer` resource created to interact with the Kopia repository of workloads running in the `monitoring` namespace looks like this:
+
+```yaml
+apiVersion: cr.kanister.io/v1alpha1
+kind: RepositoryServer
+metadata:
+   name: repository-monitoring
+   namespace: kanister
+   labels:
+      repo.kanister.io/target-namespace: monitoring
+spec:
+   location:
+      # required: storage location info
+      secret:
+         name: location
+         namespace: kanister
+      # required: creds to access the location
+      credentialSecret:
+         name: loc-creds
+         namespace: kanister
+   repository:
+      # repository must be created manually before creating server CR
+      # required: path for the repository - will be relative sub path
+      # within the path prefix specified in the location
+      rootPath: /repo/monitoring/
+      # required: password to access the repository
+      passwordSecret:
+         name: repository-monitoring-password
+         namespace: kanister
+      # required: these values will be used by the repository server controller
+      # for the connection from the server to the repository. Kopia will use
+      # these values to track and control operations
+      username: kanisterAdmin
+      hostname: monitoring.app
+   server:
+      # required: server admin details
+      adminSecret:
+         name: repository-monitoring-server-admin
+         namespace: kanister
+      # required: TLS certificate required for secure communication between the Kopia client and server
+      tlsSecret:
+         name: repository-monitoring-server-tls
+         namespace: kanister
+      # required: repository users list
+      userAccessSecret:
+         name: repository-monitoring-server-access
+         namespace: kanister
+      # required: selector for network policy required to enable cross-namespace access to the server
+      labelSelector:
```

Comment:
Can we be more explicit about this being selectors for network policy, and not e.g., `Service`? How about, following the example in the network policy [doc](https://kubernetes.io/docs/concepts/services-networking/network-policies/#networkpolicy-resource), something like:

```yaml
server:
    networkPolicy:
        namespaceSelector:
            matchLabels:
                project: myproject
        podSelector:
            matchLabels:
                role: frontend
```

Reply:
I agree with @ihcsim here. @pavannd1 we should make the `labelSelector` field more descriptive. Users might get confused about the type of these labels. They might question whether the labels here are pod-selectors or namespace-selectors. Better we explicitly ask them.

Reply:
Also makes it easier for controller to define the ingress rule in `NetworkPolicy`. Rather than making assumptions about the kind of selector these labels would be, controller would directly use these selectors while creating the `NetworkPolicy`.